Fix queries with FINAL give wrong result when table does not use adaptive granularity

canhld94 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Fix queries with FINAL give wrong result when table does not use adaptive granularity

Close #62426

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ] <!---ci_include_fast--> Fast test
- [ ] <!---ci_include_integration--> Integration Tests
- [ ] <!---ci_include_stateless--> Stateless tests
- [ ] <!---ci_include_stateful--> Stateful tests
- [ ] <!---ci_include_unit--> Unit tests
- [ ] <!---ci_include_performance--> Performance tests
- [ ] <!---ci_include_asan--> All with ASAN
- [ ] <!---ci_include_tsan--> All with TSAN
- [ ] <!---ci_include_analyzer--> All with Analyzer
- [ ] <!---ci_include_KEYWORD--> Add your option here

#### Exclude tests:
- [ ] <!---ci_exclude_fast--> Fast test
- [ ] <!---ci_exclude_integration--> Integration Tests
- [ ] <!---ci_exclude_stateless--> Stateless tests
- [ ] <!---ci_exclude_stateful--> Stateful tests
- [ ] <!---ci_exclude_performance--> Performance tests
- [ ] <!---ci_exclude_asan--> All with ASAN
- [ ] <!---ci_exclude_tsan--> All with TSAN
- [ ] <!---ci_exclude_msan--> All with MSAN
- [ ] <!---ci_exclude_ubsan--> All with UBSAN
- [ ] <!---ci_exclude_coverage--> All with Coverage
- [ ] <!---ci_exclude_aarch64--> All with Aarch64
- [ ] <!---ci_exclude_KEYWORD--> Add your option here

#### Extra options:
- [ ] <!---do_not_test--> do not test (only style check)
- [ ] <!---no_merge_commit--> disable merge-commit (no merge from master before tests)
- [ ] <!---no_ci_cache--> disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ] <!---batch_0--> 1
- [ ] <!---batch_1--> 2
- [ ] <!---batch_2--> 3
- [ ] <!---batch_3--> 4

canhld94 committed:
fix part splitter wrongly add ranges with undefined end mark value to non-intersect part

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

kitaisreal commented on tests/queries/0_stateless/03033_final_undefined_last_mark.sql:
```diff
@@ -0,0 +1,21 @@
+INSERT INTO account_test
+    SELECT * FROM (SELECT * FROM generateRandom('id UInt64, row_ver UInt64',1234) LIMIT 1000) WHERE row_ver > 14098131981223776000;
+
+SELECT 'GOOD', * FROM account_test FINAL WHERE id = 11338881281426660955 SETTINGS split_parts_ranges_into_intersecting_and_non_intersecting_final = 1;
```
It is better to add another query in test:
```
SELECT 'GOOD', * FROM account_test FINAL WHERE id = 11338881281426660955 SETTINGS split_parts_ranges_into_intersecting_and_non_intersecting_final = 0;
```
So it is easier to compare outputs.

robot-ch-test-poll commented:
<!-- automatic status comment for PR #62432 from canhld94/ClickHouse:final_undefined_last_mark -->
*This is an automated comment for commit e8f02af78c418f7c0a521bd48d49fcfb91db455f with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62432/e8f02af78c418f7c0a521bd48d49fcfb91db455f/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

canhld94 committed:
add reference query to test

Signed-off-by: Duc Canh Le <duccanh.le@ahrefs.com>

canhld94 committed:
Merge branch 'master' into final_undefined_last_mark

Try fixing CI

canhld94 commented:
AST Fuzzer test is failing, but I think it doesn't relates to the change this PR. Can reproduce in upstream `HEAD` even when turning off splitting parts https://fiddle.clickhouse.com/d5c476f8-9fb9-4ee1-bf10-0491e173bafb

canhld94 commented:
> AST Fuzzer test is failing, but I think it doesn't relates to the change this PR. Can reproduce in upstream `HEAD` even when turning off splitting parts https://fiddle.clickhouse.com/d5c476f8-9fb9-4ee1-bf10-0491e173bafb
> 
> Update: even without `FINAL`

#62741

canhld94 commented:
Stress test failure relates to  database replicated #62742

canhld94 commented:
Ping @Algunenano I think this PR is ready to merge

alexey-milovidov commented:
@canhld94, what is with these logical errors detected by our fuzzers?

canhld94 commented:
> @canhld94, what is with these logical errors detected by our fuzzers?

@alexey-milovidov this is a known bug, I opened an issue for it https://github.com/ClickHouse/ClickHouse/pull/62432 and have an attempt to fix in https://github.com/ClickHouse/ClickHouse/pull/62776 but not yet found  a good approach.

Algunenano committed:
Merge remote-tracking branch 'blessed/master' into final_undefined_last_mark

Algunenano committed:
Simplify test

Algunenano commented:
Sorry for the delay reviewing this. Initially it slipped through the cracks and then I got sidetracked by bugs in related code (but not in this PR). I've simplified the test so it doesn't depend on generateRandom and updated with master to make sure things are still ok.

canhld94 commented:
> Sorry for the delay reviewing this. Initially it slipped through the cracks and then I got sidetracked by bugs in related code (but not in this PR). I've simplified the test so it doesn't depend on generateRandom and updated with master to make sure things are still ok.

Thanks! Nevertheless, the AST fuzzer failure is still there, not causing by this PR but merging this PR will reproduce the failure in master. May be we need to push to fix this bug soon.
